docs/debug/3_api_te_calls.ipynb

```diff
@@ -0,0 +1,192 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Calls to Nvidia-DL-Framework-Inspect\n",
+    "\n",
+    "Let's look deeper into how Nvidia-DL-Framework-Inspect with Transformer Engine work together. TransformerEngine layers have some hook calls inside each of the GEMMs. Users can define feature classes or use feature classes provided with TE. File `config.yaml` describes which hooks need to be used for which layers. Nvidia-DL-Framework-Inspect combines 3 things: TE training, feature classes and `config.yaml` and takes care of inserting hooks in the correct places. This process is illustrated in the image below.\n",
+    "\n",
+    "<figure align=\"center\">\n",
+    "<img src=\"./img/api_calls1.svg\">\n",
+    "    <figcaption> Fig 1: Example of Nvidia-DL-Framework-Inspect affecting training script with 1 Linear Layer. For tensors mentioned in `config.yaml`, behavior of `modify_tensor_enabled()` and `modify_tensor()` calls are substituted with definitions from the feature class. Other calls return default values - in fact they do nothing. </figcaption>\n",
+    "</figure>\n",
+    "\n",
+    "In this page, all calls from TransformerEngine to the Nvidia-DL-Framework-Inspect for each GEMM are listed. The order of these calls is illustrated in the image below.\n",
+    "\n",
+    "\n",
+    "<figure align=\"center\">\n",
+    "<img src=\"./img/api_calls2.svg\">\n",
+    "    <figcaption> Fig 2: The calls to Nvidia-DL-Framework-Inspect done for Transformer Engine. There are 2 types of calls: GEMM calls and routing calls.</figcaption>\n",
+    "</figure>\n",
+    "\n",
+    "\n",
+    "\n",
+    "### Categories of the API calls\n",
+    "\n",
+    "There are 3 categories of API calls, each is used for different purposes:\n",
+    "\n",
+    "- GEMM calls - invoked during every GEMM, is used to process or quantize tensors and collect information about them,\n",
+    "- routing calls - invoked at the beginning of the forward pass - they indicate whether a feature is going to use `modify_tensor()`, ... etc.\n",
+    "\n",
+    "If all routing calls for the layer return `False`, then the layer is invoked in an optimized version with Transformer Engine fusions.\n",
+    "If any of the routing calls return `True`, layers are run without the fusion. It is necessary because some tensors cannot be accessed\n",
+    "if fusion happens. An important remark is that if no feature is used for the layer, then it should perform as fast as the layer without initializing `debug_api`.\n",
+    "\n",
+    "\n",
+    "### modify_tensor\n",
```

Comment:
This looks like an API doc. Rather than writing it here we should have it in Python and use autodocs.